### PR TITLE
[DEV-11456] make request chains wait appropriately for sync/async calls

### PR DESCRIPTION
**Pull request by nateshim-indico**
## Summary
Request chains with a wait argument use time.sleep() internally, rather than yielding a delay to the client to sleep synchronously or asynchronously as appropriate. Moved sleeping to client and making all wait calls yield a float or integer denoting sleep time

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] Jira story/task is put into PR title
- [x] Code has been self-reviewed
- [x] Hard-to-understand areas have been commented
- [x] Documentation has been updated
- [x] My changes generate no new warnings
- [x] Unit tests have been added for base functionality and known edge cases
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: indico/client/client.py [hunk 1]
```diff
@@ -47,7 +50,8 @@ def _handle_request_chain(self, chain: RequestChain):
             elif isinstance(request, RequestChain):
                 response = self._handle_request_chain(request)
                 chain.previous = response
-
+            elif isinstance(request, (float, int)):
```
Comment (pastap) [inline indico/client/client.py:53]: this might be... a little brittle if the return type is ever a float or int. Have you thought about using something like a "Backoff" class instead? really simple
```
class Backoff:
  def __init__(self, time):
    self.time = time

return Backoff(5)
```
and then you can
```
if isinstance(req, Backoff):
  sleep(Backoff.time)
```

Comment (pastap) [reply indico/client/client.py:53]: or even just leveraging a function to call like
if isinstance(request, Callable): request()...

Comment (pastap) [reply indico/client/client.py:53]: hell yeah, sounds good - for sleeps, could just... 0 debounce or something yeah - great call

File: indico/queries/datasets.py [hunk 2]
```diff
@@ -215,6 +215,7 @@ def __init__(
         name: str,
         files: List[str],
         wait: bool = True,
+        max_wait_time: Tuple[int, float] = 5,
```
Comment (nateshim-indico) [inline indico/queries/datasets.py:218]: Is it okay to add here or should this be at bottom of kwarg list

File: indico/client/request.py [hunk 3]
```diff
@@ -90,13 +90,13 @@ def requests(self):
 
 
 class Debouncer:
-    def __init__(self, max_timeout: int = 5):
+    def __init__(self, max_timeout: Tuple[int, float] = 5):
```
Comment (nateshim-indico) [inline indico/client/request.py:93]: Added kwarg `max_wait_time` to all client calls that use `Debouncer` to allow access to change this variable.

File: indico/client/request.py [hunk 4]
```diff
@@ -90,13 +90,13 @@ def requests(self):
...
         self.timeout = 0
         self.max_timeout = max_timeout or 5  # prevent None and 0
 
-    def backoff(self):
+    def backoff(self) -> Tuple[int, float]:
```
Comment (pastap) [inline indico/client/request.py:97]: same Union comment here

File: indico/queries/datasets.py (hunk 2 above)
Comment (pastap) [reply indico/queries/datasets.py:218]: good call - bottom of list - just in case we got some folks who blindly pass params

Comment (pastap) [reply indico/queries/datasets.py:218]: and same Union, not Tuple comment

File: indico/queries/export.py [hunk 5]
```diff
@@ -54,7 +54,7 @@ def __init__(
         frozen_labelset_ids: List[int] = None,
         combine_labels: LabelResolutionStrategy = LabelResolutionStrategy.ALL.name,
         file_info: bool = None,
-        anonymoous: bool = None,
+        anonymous: bool = None,
```
Comment (pastap) [inline indico/queries/export.py:57]: lmao - I'm hoping no one used this otherwise technically a backwards-incompat change. will let @goatrocks make the final call here

File: indico/queries/jobs.py [hunk 6]
```diff
@@ -67,11 +68,17 @@ class JobStatus(RequestChain):
 
     previous: Job = None
 
-    def __init__(self, id: str, wait: bool = True, request_interval=0.2, timeout=None):
+    def __init__(
+        self,
+        id: str,
+        wait: bool = True,
+        timeout: Tuple[int, float] = None,
+        max_wait_time: Tuple[int, float] = 5,
```
Comment (pastap) [inline indico/queries/jobs.py:76]: what's the diff? (also Union)

Comment (pastap) [reply indico/queries/jobs.py:76]: also - docstring update?

File: indico/queries/model_groups/model_groups.py [hunk 7]
```diff
@@ -27,16 +23,19 @@ class GetModelGroup(RequestChain):
         ModelGroup object
     """
 
-    def __init__(self, id: int, wait: bool = False):
+    def __init__(
+        self, id: int, wait: bool = False, max_wait_time: Tuple[int, float] = 5
+    ):
         self.id = id
         self.wait = wait
+        self.max_wait_time = max_wait_time
```
Comment (pastap) [inline indico/queries/model_groups/model_groups.py:31]: docstring update, Union instead of Tuple, etc

File: indico/queries/export.py (hunk 5 above)
Comment (goatrocks) [reply indico/queries/export.py:57]: If we call it out clearly in the release notes, given this isn't something like the submission api... should be ok to fix 🐄  😨

File: indico/queries/jobs.py (hunk 6 above)
Comment (nateshim-indico) [reply indico/queries/jobs.py]: timeout makes call throw `IndicoTimeoutError` after a specified period of time. `max_wait_time` denotes the max wait time per retry call until JobStatus is COMPLETE. 

Updated docstring to hopefully make the distinction clear

File: indico/queries/export.py (hunk 5 above)
Comment (nateshim-indico) [reply indico/queries/export.py:57]: Found this library that support argument deprecation: https://docs.twisted.org/en/stable/api/twisted.python.deprecate.html

Don't know how worth it would be to add it for this

File: indico/queries/datasets.py [hunk 8]
```diff
@@ -552,21 +559,22 @@ def __init__(
...
         self.wait = wait
+        self.max_wait_time = max_wait_time
 
     def requests(self):
         yield _ProcessFiles(self.dataset_id, self.datafile_ids)
-        debouncer = Debouncer()
         yield GetDatasetFileStatus(id=self.dataset_id)
         if self.wait:
             while not all(
                 f.status in ["PROCESSED", "FAILED"] for f in self.previous.files
```
Comment (pastap) [inline indico/queries/datasets.py:574]: dumb question - does self.previous pick up the last `yield` in which case it'd mistakenly pick up the debouncer value instead of getdatasetfilestatus?

File: indico/queries/submission.py [hunk 9]
```diff
@@ -378,12 +378,14 @@ def __init__(
         check_status: str = None,
         wait: bool = False,
         timeout: Union[int, float] = 30,
+        max_wait_time: Tuple[int, float] = 5,
```
Comment (pastap) [inline indico/queries/submission.py:381]: Union

File: indico/queries/datasets.py [hunk 10]
```diff
@@ -1,13 +1,13 @@
 # -*- coding: utf-8 -*-
 
 import json
-import jsons
 import tempfile
 from pathlib import Path
-from typing import List, Union, Dict, Optional
+from typing import Dict, List, Optional, Tuple, Union
```
Comment (pastap) [inline indico/queries/datasets.py:6]: unneeded Tuple import

File: indico/queries/export.py (hunk 5 above)
Comment (pastap) [reply indico/queries/export.py:57]: or just print a `warnings.warn` if someone uses the old arg?

Comment (goatrocks) [reply indico/queries/export.py:57]: warnings.warn is good -- we have a deprecation library already but unfortunately it doesn't support params (or it didn't when it was added). :(

File: indico/queries/jobs.py (hunk 6 above)
Comment (goatrocks) [reply indico/queries/jobs.py]: `retry_time_interval` or `time_between_retries` would be a clearer name for this variable I think.

File: indico/queries/submission.py (hunk 9 above)
Comment (nateshim-indico) [reply indico/queries/submission.py:381]: smh mb * 5

File: indico/queries/datasets.py (hunk 8 above)
Comment (nateshim-indico) [reply indico/queries/datasets.py:574]: just tested and it does not. I did not think of that tbh so good thing for me to be aware of going forward thanks

File: indico/queries/jobs.py [hunk 11]
```diff
@@ -67,15 +69,22 @@ class JobStatus(RequestChain):
...
-    def __init__(self, id: str, wait: bool = True, request_interval=0.2, timeout=None):
+    def __init__(
+        self,
+        id: str,
+        wait: bool = True,
+        timeout: Union[int, float] = None,
+    ):
         self.id = id
         self.wait = wait
-        self.request_interval = request_interval
```
Comment (nateshim-indico) [inline indico/queries/jobs.py:73]: @goatrocks @pastap  Should we `warnings.warn` before removing?

File: indico/queries/export.py (hunk 5 above)
Comment (nateshim-indico) [reply indico/queries/export.py:60]: added typo argument back and checks for it as discussed

File: indico/queries/jobs.py (hunk 6 above)
Comment (nateshim-indico) [reply indico/queries/jobs.py]: changed `max_wait_time` to `request_interval` to match existing argument in `JobStatus` query

**Review by jacobmanderson — APPROVED** [review]

File: indico/queries/datasets.py [hunk 12]
```diff
@@ -196,12 +196,17 @@ class CreateDataset(RequestChain):
...
+        files (List[str]): List of path names to the dataset files.
+        wait (bool, optional): Wait for the dataset to upload and finish. Defaults to True.
+        dataset_type (str, optional): Type of dataset to create [TEXT, DOCUMENT, IMAGE]. Defaults to TEXT.
+        from_local_images (bool, optional): Flag whether files are local images or not. Defaults to False.
+        image_filename_col (str, optional): Image filename column. Defaults to 'filename'.
+        batch_size (int, optional): Size of file batch to upload at a time. Defaults to 20.
+        ocr_engine (OcrEngine, optional): Specify an OCR engine [OMNIPAGE, READAPI, READAPI_V2, READAPI_TABLES_V1]. Defaults to None.
+        omnipage_ocr_options (OmnipageOcrOptionsInput, optional): If using Omnipage, specify Omnipage OCR options. Defaults to None.
+        read_api_ocr_options: (ReadApiOcrOptionsInput, optional): If using ReadAPI, specify ReadAPI OCR options. Defaults to None.
+        request_interval (int or float, optional): The maximum time in between retry calls when waiting. Defaults to 5.
```
Comment (goatrocks) [inline indico/queries/datasets.py:209]: bless your doc updates :)

File: indico/queries/datasets.py [hunk 13]
```diff
@@ -538,9 +544,10 @@ class ProcessFiles(RequestChain):
     Process files associated with a dataset and add corresponding data to the dataset
 
     Args:
-        dataset_id (int): ID of the dataset
-        datafile_ids (List[str]): IDs of the datafiles to process
-        wait (bool): Block while polling for status of files
+        dataset_id (int): ID of the dataset.
+        datafile_ids (List[str]): IDs of the datafiles to process.
+        wait (bool, optional): Block while polling for status of files. Defaults to True.
+        request_interval (int or float, optional): The maximum time in between retry calls when waiting. Defaults to 5.
```
Comment (goatrocks) [inline indico/queries/datasets.py:550]: 5 seconds? Minutes?

File: indico/queries/export.py (hunk 5 above)
Comment (goatrocks) [reply indico/queries/export.py:60]: Thank you Nate!

**Review by goatrocks — DISMISSED** [review]
lgtm pending @mawelborn 's feedback 👍

**Review by mawelborn — CHANGES_REQUESTED** [review]
Couple issues:

1. [Debouncing](https://underscorejs.org/#debounce) is a specific event-based programming technique that I was surprised to see pop up here. Reading the code, I see it's actually doing something related to backoff.
2. Regardless of the `request_interval` or `Debouncer.max_timeout` specified, the client will always sleep 1 second because that's where the `debouncer` starts, and `debouncer.backoff()` never gets called twice on a single instance.

As someone who creates request chains, but isn't overly familiar with the client codebase, I found this to be a bit confusing. If the desire is to avoid yielding a bare `float` for delays (understandable), my recommendation would be to replace it with a simple `Delay` dataclass to improve readability and fix the issue of not being able to yield a longer delay.

```python
@dataclass
class Delay:
    seconds: float
```

```python
class IndicoClient:
    ...

    def _handle_request_chain(self, chain: RequestChain):
        ...
        elif isinstance(request_or_delay, Delay):
            time.sleep(request_or_delay.seconds):
```

```python
class WaitUntilComplete(RequestChain):
    ...

    def requests(self) -> Iterator[GraphQLRequest | Delay]:
        yield GetSubmission(self.submission_id)
        while self.previous.status != "COMPLETE":
            yield Delay(seconds=8)
            yield GetSubmission(self.submission_id)
```

**Comment by nateshim-indico** [comment]
Debouncer was used for CreateExport call (I don't know the history of why) and so we leveraged it throughout for consistency.

@goatrocks @pastap I'm thinking Debouncer is probably not needed for CreateExport and we can use simple Delay class throughout instead? I don't think it makes sense to have both especially if Debouncer will be used for a single call (CreateExport)

**Review by mawelborn — APPROVED** [review]
Beautiful!

**Review by goatrocks — APPROVED** [review]
